pallets/pallet-lottery/src/lib.rs

```diff
@@ -0,0 +1,1077 @@
+                UnstakingCollators::<T>::put(unstaking);
+            }
+        }
+
+        #[named]
+        fn do_rebalance_remaining_funds() -> DispatchResult {
+            log::trace!(function_name!());
+            // Only restake what isn't needed to service outstanding withdrawal requests
+            let stakable_balance = Self::current_prize_pool();
+            if stakable_balance.is_zero() {
+                log::debug!("Nothing to restake");
+                return Ok(());
+            }
+            for (collator, amount_to_stake) in
+                Self::calculate_deposit_distribution(stakable_balance)
+            {
+                Self::do_stake_one_collator(collator.clone(), amount_to_stake)?;
+                TotalPot::<T>::mutate(|balance| *balance += amount_to_stake);
+                log::debug!(
+                    "Rebalanced {:?} to collator {:?}",
+                    amount_to_stake,
+                    collator
+                );
+            }
+            Ok(())
+        }
+
+        /// This fn schedules a single shot payout of all matured withdrawals
+        /// Main usage: Automatic execution in the course of a drawing
+        /// It can also be manually invoke by T::ManageOrigin to reprocess withdrawals that
+        /// previously failed, e.g. due to the pallet running out of gas funds
+        /// A withdrawal is considered "matured" if its staking timelock expired
+        #[named]
+        fn do_process_matured_withdrawals() -> DispatchResult {
+            log::trace!(function_name!());
+            if <WithdrawalRequestQueue<T>>::get().is_empty() {
+                return Ok(()); // nothing to do
+            }
+            let funds_available_to_withdraw = Self::surplus_funds();
+            log::debug!(
+                "Withdrawable funds: {:?}",
+                funds_available_to_withdraw.clone()
+            );
+            if funds_available_to_withdraw.is_zero() {
+                return Ok(()); // nothing to do
+            }
+            let now = <frame_system::Pallet<T>>::block_number();
+            // Pay down the list from top (oldest) to bottom until we've paid out everyone or run out of available funds
+            <WithdrawalRequestQueue<T>>::mutate(|request_vec| -> Result<(), DispatchError> {
+                let mut left_overs: Vec<Request<_, _, _>> = Vec::new();
+                for request in request_vec.iter() {
+                    // Don't pay anyone still timelocked
+                    if now < request.block + <T as Config>::UnstakeLockTime::get() {
+                        left_overs.push((*request).clone());
+                        continue;
+                    }
+                    // stop paying people if we've run out of free funds.
+                    // The assumption is the collators serving these requests will
+                    // finish unstaking next round ( next lottery drawing )
+                    if request.balance > funds_available_to_withdraw {
```

Comment:
we should update funds_available_to_withdraw after each successful transfer below

Reply:
thanks for this good catch. this was a fairly large oversight. moved available_balance into the loop